tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift

```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    let hookAddr = unsafeBitCast(GetProcAddress(hKernel32, "LoadLibraryW"),
+                                 to: LPTHREAD_START_ROUTINE.self)
+    let hThread: HANDLE = CreateRemoteThread(self.process, nil, 0, hookAddr,
+                                             dllPathRemote, 0, nil)
+    if hThread == HANDLE(bitPattern: 0) {
+      print("CreateRemoteThread failed \(GetLastError())")
+      return
+    }
+    defer { CloseHandle(hThread) }
+
+    // The main heap iteration loop.
+    while true {
+      let wait = WaitForSingleObject(hReadEvent, waitTimeoutMs)
+      if wait != WAIT_OBJECT_0 {
+        print("WaitForSingleObject failed \(wait)")
+        return
+      }
+
+      let entryCount = bufSize/MemoryLayout<(UInt, UInt)>.size
+      let buffer = buf.bindMemory(to: (UInt, UInt).self, capacity: 1)
+      let (address0, _) = buffer.pointee
+      if address0 == UInt.max {
+        // The buffer containing zero entries, indicated by the first entry
+        // contains -1, means, we are done. Break out of loop.
+        if !SetEvent(hWriteEvent) {
+          print("SetEvent failed \(GetLastError())")
+          return
+        }
+        break
+      }
+      for i in 0..<entryCount {
+        let (address, size) = buffer.advanced(by: i).pointee
+        if address == 0 {
+          // Done. Break out of loop.
+          break
+        }
+        body(swift_addr_t(address), UInt64(size))
+      }
+
+      if !SetEvent(hWriteEvent) {
+        print("SetEvent failed \(GetLastError())")
+        return
       }
+    }
 
-      repeat {
-        if entry.dwFlags & DWORD(LF32_FREE) == DWORD(LF32_FREE) { continue }
-        body(swift_addr_t(entry.dwAddress), UInt64(entry.dwBlockSize))
-      } while Heap32Next(&entry)
+    let wait = WaitForSingleObject(hThread, waitTimeoutMs)
+    if wait != WAIT_OBJECT_0 {
+      print("WaitForSingleObject on LoadLibrary failed \(wait)")
+      return
+    }
+
+    var threadExitCode: DWORD = 0
+    GetExitCodeThread(hThread, &threadExitCode)
+    if threadExitCode == 0 {
+      print("LoadLibraryW failed \(threadExitCode)")
+      return
+    }
+
+    // Unload the dll and deallocate the dll path from the remote process
+    if !unloadDllAndPathRemote(dwProcessId: dwProcessId, hKernel32: hKernel32,
+                               dllPathRemote: dllPathRemote) {
+      print("Failed to unload the remote dll")
+      return
+    }
+  }
+
+  private func allocateDllPathRemote() -> UnsafeMutableRawPointer? {
+    // The path to the dll assuming it's in the same directory as swift-inspect.
+    let swiftInspectPath = ProcessInfo.processInfo.arguments[0]
+    let dllPath: String =
+        URL(fileURLWithPath: swiftInspectPath).deletingLastPathComponent().path +
+        "\\SwiftInspectClient.dll"
+    var dllPathRemote = UnsafeMutableRawPointer(bitPattern: 0)
+    var dllPathUtf16Len: UInt = 0
+    var dllPathWritten = false
+    dllPath.withCString(encodedAs: UTF16.self) {
```

Comment:
This is unsafe.  We have a path limitation to 261 characters.  We absolutely need to do this inside the closure to ensure that we have a NT-style path ...


```swift
URL(fileURLWithPath: swiftInspectPath)
        .deletingLastPathComponent()
        .appendingPathComponent("SwiftInspectClient.dll")
        .withUnsafeFileSystemRepresentation {
  #"\\?\\#(String(decodingCString: $0!, as: UTF16.self))"#.withCString(encodedAs: UTF16.self) {
    let faAttributes: WIN32_FILE_ATTRIBUTE_DATA = .init()
    guard GetFileAttributesExW($0, GetFileExInfoStandard, &faAttributes),
          faAttributes.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT == 0 else {
      return nil
    }

    let szLength = SIZE_T(_wcslen($0) + 1)
    guard let allocation = VirtualAllocEx(self.process, nil, szLength,
                                          DWORD(MEM_COMIT), DWORD(PAGE_READWRITE)) else {
      return nil
    }

    if !WriteProcessMemory(self.process, allocation, $0, szLength, nil) {
      VirtualFreeEx(self.process, allocation, 0, DWORD(MEM_RELEASE)
      return nil
    }

    return allocation
  }
}
```

Reply:
Done. Can you elaborate? I understand there's a path limitation. What's a NT-style path? And how does using a closure ensure it? Or what did the original code do wrong?

Reply:
The `\\?\` prefix creates a NT-style path.  It is a path that bypasses the Win32 layer and goes into the kernel namespace, which allows it to bypass the 261 character limitation.  The closure doesn't do that, its the interior string construction which does that.  The problem is that the path must be normalised first as the normalization is something that the Win32 layer normally handles.  Due to the memory handling in Foundation, we need to use a closure (`withUnsafeFileSystemRepresentation`).  That mostly normalizes the path, and then we can create the long path.